projects/vdk-plugins/vdk-data-sources/src/vdk/plugin/data_sources/data_source.py

```diff
@@ -0,0 +1,245 @@
+    destination_table: Optional[str] = None
+
+
+class StopDataSourceStream(Exception):
+    """Signal the end of a stream and there's no more data"""
+
+
+class RetryDataSourceStream(Exception):
+    """Signal the stream ingestion should be retried"""
+
+
+class IDataSourceStream:
+    """
+    Abstract class for a Data Source Stream, representing a channel or resource to read data.
+
+    For example each table in a database could be a stream.
+    Or each separate API endpoint for an API could be a stream and so on.
+
+    :Example::
+
+        class MyStream(IDataSourceStream):
+            def name(self):
+                return "My Stream"
+
+            def read(self):
+                yield DataSourcePayload(...)
+
+    """
+
+    @abstractmethod
+    def name(self) -> str:
+        """
+        :return: unique (within the data souce) name of the data source stream which can be used as a key and find by it.
+        """
+        pass
+
+    @abstractmethod
+    def read(self) -> Iterable[DataSourcePayload]:
+        """
+        Generator method or Iterator for reading data from the stream.
+
+        :return: An iterable of DataSourcePayload objects.
+
+        """
+        pass
+
+
+class IDataSource:
+    """
+    Abstract class for a Data Source, responsible for managing the connection and providing data streams.
+
+    :Example::
+
+        class MyDataSource(IDataSource):
+            def connect(self, config, state):
+                pass
+
+            def disconnect(self):
+                pass
+
+            def streams(self):
+                return [MyStream()]
+
+    """
+
+    @abstractmethod
+    def configure(self, config: IDataSourceConfiguration):
+        """
+        Configure the data source.
+
+        :param config: Data source configuration object.
+        """
+
+    @abstractmethod
+    def connect(self, state: Optional[IDataSourceState]):
+        """
+        Establish a connection using provided configuration and last saved state.
+
+        :param state: Data source state object.
+        """
+
+    @abstractmethod
+    def disconnect(self):
+        """
+        Disconnect and clean up resources if needed.
+        """
+
+    @abstractmethod
+    def streams(self) -> List[IDataSourceStream]:
+        """
+        Get the available streams for this data source.
+
+        :return: List of IDataSourceStream objects.
+        """
+
+
+class DataSourcesAggregatedException(Exception):
+    """
+    Exception to aggregate multiple Data Sources failures.
+
+    :Example::
+
+        DataSourcesAggregatedException({"Stream1": Exception1, "Stream2": Exception2})
+    """
+
+    def __init__(self, data_streams_exceptions: Dict[str, Exception]):
+        super().__init__(
+            f"Data Sources failed to ingest data: {data_streams_exceptions}"
+        )
+        self.data_streams_exceptions = data_streams_exceptions
+
+
+@dataclass
+class DataSourceError:
```

Comment:
why did you not call it an exception? 
I would think that error is more serious. 

like out of memory error.